### PR TITLE
Select: Preserve value when `allowCustomValue` is set

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- improves some typings in `SelectBase`
- adds support to `<Select>` for editing a custom value 
  - this only applies when `allowCustomValue` is true and `isMulti` is false
  - this works by making the react-select input state controlled by us
    - `onMenuOpen` we restore the internal input state to be the selected values label
    - `onMenuClose` we clear the input state else `react-select` doesn't show the `SingleValue` component correctly
    - `onInputChange` we update the internal input value state
- updates some unit tests that relied on the old behaviour

**Why do we need this feature?**

- so users can edit a custom value once entered

**Who is this feature for?**

- anyone using adhoc filters!

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/hyperion-planning/issues/27

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Select/CustomInput.tsx [hunk 1]
```diff
@@ -13,3 +13,7 @@ export const CustomInput = (props: InputProps) => {
 
   return <components.Input {...props} data-testid={testId} />;
 };
+
+export const CustomInputAlwaysVisible = (props: InputProps) => {
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Select/CustomInput.tsx:17]: lol

File: packages/grafana-prometheus/src/querybuilder/components/PromQueryBuilder.test.tsx [hunk 2]
```diff
@@ -174,35 +174,20 @@ describe('PromQueryBuilder', () => {
...
       labels: [],
       operations: [],
     });
-    await openMetricSelect(container);
-    await userEvent.click(screen.getByText('histogram_metric_bucket'));
     await waitFor(() => expect(screen.getByText('hint: add histogram_quantile')).toBeInTheDocument());
   });
 
   it('shows hints for counter metrics', async () => {
-    const { container } = setup({
```
Comment (ashharrison90) [inline packages/grafana-prometheus/src/querybuilder/components/PromQueryBuilder.test.tsx:188]: this test was duplicated below, so removed it

File: public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx [hunk 3]
```diff
@@ -111,6 +111,7 @@ describe('useCreatableSelectPersistedBehaviour', () => {
     await userEvent.click(input);
 
     // we expect 2 elemnts having "Option 2": the input itself and the option.
```
Comment (matyax) [inline public/app/plugins/datasource/elasticsearch/components/hooks/useCreatableSelectPersistedBehaviour.test.tsx:113]:
```suggestion
    // we expect 2 elements having "Option 2": the input itself and the option.
```

**Review by matyax — COMMENTED** [review]
👍  logs changes.

**Review by joshhunt — COMMENTED** [review]
I think this is secretly actually a very big an impactful change... it's a pretty big conceptual change to how custom values work and users interact with them. In saying that, I'm pretty supportive of this - this half-answers a question about fixing this behaviour in the new select component.

Two issues I spotted with it:

When editing a value, trying to click to place the cursor doesn't work. It just closes the dropdown. The only way to move the cursor is to either backspace or use arrow keys.

I think it's only used in plugins, but conincidentally I noticed a style regression with multi select with the `noMultiValueWrap` prop enabled.

image


https://developers.grafana.com/ui/canary/index.html?path=/story/forms-select--multi-plain-value&args=noMultiValueWrap:!true

vs 

http://localhost:9001/?path=/story/forms-select--multi-plain-value&args=noMultiValueWrap:!true

**Comment by ashharrison90** [comment]
> When editing a value, trying to click to place the cursor doesn't work. It just closes the dropdown. The only way to move the cursor is to either backspace or use arrow keys.

yeah i haven't quite find a way to fix that yet. there is one way, but it involves setting the `controlShouldRenderValue` prop to false, which significantly alters the behaviour for unit tests (basically the `SingleValue` component is never rendered and we instead rely on the internal `<input>` showing the value)

> I think it's only used in plugins, but conincidentally I noticed a style regression with multi select with the noMultiValueWrap prop enabled.

i see the same behaviour in canary as on this branch? but not on https://developers.grafana.com/ui/latest/index.html?path=/story/forms-select--multi-plain-value&args=noMultiValueWrap:!true

so must be something that's regressed recently. will investigate that separately.

**Comment by ashharrison90** [comment]
> I think it's only used in plugins, but conincidentally I noticed a style regression with multi select with the noMultiValueWrap prop enabled.

yep, it's caused by https://github.com/grafana/grafana/pull/87085 😬 fyi @kaydelaney 
you can repro by going to the `Select` -> `Multi plain value` story, setting `noMultiValueWrap` to true and adding options. see https://developers.grafana.com/ui/latest/index.html?path=/story/forms-select--multi-plain-value for what it should look like

**Comment by ashharrison90** [comment]
> When editing a value, trying to click to place the cursor doesn't work. It just closes the dropdown. The only way to move the cursor is to either backspace or use arrow keys.

ok, dug a bit more into it this morning. it's caused by the `menuShouldBlockScroll` option passed to `react-select`. here's some options, roughly in order of preference:
- do nothing. it's the current behaviour of `Select` anyway. i guess this exposes the problem a little more, but still 🤷 
- set `menuShouldBlockScroll=false` for `Select`s that allow custom values. downside is it introduces an inconsistency between different types of `Select` where one can scroll and the other can't. and it introduces the downsides explained below for those `Select`s
- set `menuShouldBlockScroll=false` for all `Select`s. this can cause some slightly weird issues where you scroll a select outside of the scroll container but the menu is still visible and overlaying everything, e.g. panel edit:
image

**Comment by dprokop** [comment]
@ashharrison90 i think it's ok to "do nothing. it's the current behaviour of Select anyway. i guess this exposes the problem a little more, but still 🤷" :) I mean - i personally did not find this problematic, thinking about the following proposals, they bring to much of a change to the UX imo.

**Comment by joshhunt** [comment]
I would have hoped we could listen to the click even on the input, handle it there, and prevent it from propagating up to the rest of react-select?

Otherwise, I think it's fine (but not ideal) to leave the behaviour as-is then. I think this highlights the existing awkward behaviour even more, but it's probably fine to keep it around for a bit more. 👍

**Comment by joshhunt** [comment]
also good catch on the weird display issue - when I spotted it it wasn't on canary yet, but I see it now is 👍

**Review by joshhunt — APPROVED** [review]

**Review by dprokop — APPROVED** [review]

File: packages/grafana-ui/src/components/Select/SelectBase.tsx [hunk 4]
```diff
@@ -268,12 +275,41 @@ export function SelectBase<T, Rest = {}>({
...
     creatableProps.allowCreateWhileLoading = allowCreateWhileLoading;
     creatableProps.formatCreateLabel = formatCreateLabel ?? defaultFormatCreateLabel;
     creatableProps.onCreateOption = onCreateOption;
     creatableProps.createOptionPosition = createOptionPosition;
     creatableProps.isValidNewOption = isValidNewOption;
+
+    // code needed to allow editing a custom value once entered
+    // we only want to do this for single selects, not multi
+    if (!isMulti) {
+      creatableProps.inputValue = internalInputValue;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Select/SelectBase.tsx:288]: @ashharrison90 

what of passed inputValue via props?

File: packages/grafana-ui/src/components/Select/SelectBase.tsx [hunk 5]
```diff
@@ -268,12 +275,41 @@ export function SelectBase<T, Rest = {}>({
...
+      creatableProps.onMenuOpen = () => {
+        // make sure we call the base onMenuOpen if it exists
+        commonSelectProps.onMenuOpen?.();
+
+        // restore the input state to the selected value
+        setInternalInputValue(selectedValue?.[0]?.label ?? '');
+      };
+      creatableProps.onInputChange = (val, actionMeta) => {
+        // make sure we call the base onInputChange
+        commonSelectProps.onInputChange(val, actionMeta);
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Select/SelectBase.tsx:298]: @ashharrison90
 
you should return what onInputChange returns here as that is important (controls some behaviors)

**Pull request merged**